### PR TITLE
Ui/TRA-4607 - replay mizu requests

### PR DESCRIPTION
**Pull request by leon-up9**
Added replay option:
image

image

### CODE REVIEW

File: ui-common/src/components/EntryDetailed/EntrySections/EntrySections.tsx [hunk 1]
```diff
@@ -117,6 +117,37 @@ interface EntryBodySectionProps {
...
+            });
+        } else if (protobufFormats.some(format => contentType?.indexOf(format) > -1)) {
+            // Replace all non printable characters (ASCII)
+            const protobufDecoder = new ProtobufDecoder(bodyBuf, true);
+            const protobufDecoded = protobufDecoder.decode().toSimple();
+            if (!isPretty) return JSON.stringify(protobufDecoded);
+            return jsonBeautify(protobufDecoded, null, 2, 80);
+        }
+    } catch (error) {
+        console.log(error)
```
Comment (lirazyehezkel) [inline ui-common/src/components/EntryDetailed/EntrySections/EntrySections.tsx:146]: console.error?

File: ui-common/src/components/EntryDetailed/EntrySections/EntrySections.tsx [hunk 2]
```diff
@@ -117,6 +117,37 @@ interface EntryBodySectionProps {
     selector?: string,
 }
 
+export const formatRequest = (body: any, contentType: string, decodeBase64: boolean = true, isBase64Encoding: boolean = false, isPretty: boolean = true): string => {
+    if (!decodeBase64 || !body) return body;
+
+    const chunk = body.slice(0, MAXIMUM_BYTES_TO_FORMAT);
+    const bodyBuf = isBase64Encoding ? atob(chunk) : chunk;
+
+    try {
```
Comment (lirazyehezkel) [inline ui-common/src/components/EntryDetailed/EntrySections/EntrySections.tsx:126]: I wonder if we need the doubled try-catch both here and in the formatTextBody func..
Make sure we don't print the error twice when there is an error

File: ui-common/src/components/TrafficViewer/TrafficViewer.tsx [hunk 3]
```diff
@@ -276,8 +277,8 @@ const TrafficViewerContainer: React.FC<TrafficViewerProps> = ({
                     closeOnClick
                     rtl={false}
                     pauseOnFocusLoss
-                    draggable
-                    pauseOnHover/>
+      draggable />
```
Comment (lirazyehezkel) [inline ui-common/src/components/TrafficViewer/TrafficViewer.tsx:280]: fix indentention

File: ui-common/src/components/UI/CodeEditor/CodeEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        code,
+        onChange,
+        isDisabled = false,
+        className,
+        name,
+        variableHeight = false,
+        language = 'python',
+        errorMessage,
+        hideTooltip = false,
+        hideGutter = false
```
Comment (lirazyehezkel) [inline ui-common/src/components/UI/CodeEditor/CodeEditor.tsx:49]: it looks like you don't use most of the props

File: ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+type Row = { key: string, value: string }
+
+const KeyValueTable: React.FC<KeyValueTableProps> = ({ data, onDataChange, keyPlaceholder, valuePlaceholder }) => {
+
+    const [keyValueData, setKeyValueData] = useState([])
+
+    useEffect(() => {
+        if (!data) return;
+        let currentState = [...data, { key: "", value: "" }]
```
Comment (lirazyehezkel) [inline ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx:23]: const

File: ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        setKeyValueData(currentState)
```
Comment (lirazyehezkel) [inline ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx:24]: can we add this to the deps array and remove the disable lint?

File: ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    onClose: () => void;
+}
+
+enum RequestTabs {
+    Params = "params",
+    Headers = "headers",
+    Body = "body"
+}
+
+const isJson = (str) => {
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx:50]: move to helpers

File: ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        JSON.parse(str);
+    } catch (e) {
+        return false;
+    }
+    return true;
+}
+
+const httpMethods = ["get", "post", "put", "head", "options", "delete"]
+const TABS = [{ tab: RequestTabs.Headers }, { tab: RequestTabs.Params }, { tab: RequestTabs.Body }];
+const convertParamsToArr = (paramsObj) => Object.entries(paramsObj).map(([key, value]) => { return { key, value } })
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx:61]: add line space above and under this func

File: ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+const isJson = (str) => {
+    try {
+        JSON.parse(str);
+    } catch (e) {
+        return false;
+    }
+    return true;
+}
+
+const httpMethods = ["get", "post", "put", "head", "options", "delete"]
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx:59]: HTTP_METHODS

File: ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+    }, [debouncedPath])
+
+    useEffect(() => {
+        const newParams = getQueryStringParams(debouncedPath);
+        setParams(convertParamsToArr(newParams))
+    }, [debouncedPath])
+
+    const onModalClose = () => {
+        setRequestExpanded(true)
```
Comment (lirazyehezkel) [inline ui-common/src/components/modals/ReplayRequestModal/ReplayRequestModal.tsx:118]: why do you need to open the request and response accordions onModalClose?

File: ui-common/src/hooks/useDebounce.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { useState, useEffect } from "react";
+
+const useDebounce = (value, delay) => {
```
Comment (lirazyehezkel) [inline ui-common/src/hooks/useDebounce.ts:3]: haven't u added already this hook?
Or it was only for the enterprise?

File: ui-common/src/recoil/serviceMapModalOpen/atom.ts [hunk 12]
```diff
@@ -1,8 +0,0 @@
-import { atom } from "recoil"
-
-const serviceMapModalOpenAtom = atom({
```
Comment (lirazyehezkel) [inline ui-common/src/recoil/serviceMapModalOpen/atom.ts:3]: why it was removed?

**Review by lirazyehezkel — COMMENTED** [review]
Added some comments, pls take a look
and pay attention to the eslint warnings as well

**Comment by leon-up9** [comment]
> why it was removed? 

wasnt used.
used in Enterprise

**Comment by leon-up9** [comment]
> why do you need to open the request and response accordions onModalClose? 

i tought its good idea to reset the state bedore the next opening (which have the same state)

File: ui-common/src/components/UI/CodeEditor/CodeEditor.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    code: string,
+    name?: string,
+    onChange?: (code: string) => void,
+    isDisabled?: boolean,
+    className?: string,
+    variableHeight?: boolean,
+    language?: string,
+    errorMessage?: string,
+    hideTooltip?: boolean,
+    hideGutter?: boolean
```
Comment (lirazyehezkel) [inline ui-common/src/components/UI/CodeEditor/CodeEditor.tsx:35]: remove the unrelevant props from here as well

File: ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+const KeyValueTable: React.FC<KeyValueTableProps> = ({ data, onDataChange, keyPlaceholder, valuePlaceholder }) => {
+
+    const [keyValueData, setKeyValueData] = useState([] as Row[])
+
+    useEffect(() => {
+        if (!data) return;
+        const currentState = [...data, { key: "", value: "" }]
+        setKeyValueData(currentState)
+    }, [data])
```
Comment (lirazyehezkel) [inline ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx:25]: I saw you resolved the last previous comment but didn't add this "setKeyValueData" to the deps array.
why does it not cause an eslint warning?

Comment (leon-up9) [reply ui-common/src/components/UI/KeyValueTable/KeyValueTable.tsx:25]: i dont know really

**Review by lirazyehezkel — APPROVED** [review]

**Review by lirazyehezkel — APPROVED** [review]

**Pull request merged**